boot/freeldr/bootsect/ext.S

```diff
@@ -0,0 +1,490 @@
+
+    /* Calculate the Inode index offset */
+    mov bx, offset ExtReadInodeIndex - start
+    mov si, offset ExtReadInodeIndexBlock - start
+    mov di, offset ExtReadInodeIndexOffset - start
+    call ExtCalculateOffset
+
+    /* Read group descriptor */
+    call ExtReadGroupDescriptor
+
+    /* Set the offset address */
+    mov si, word ptr BP_REL(ExtReadInodeGroupOffset)
+
+    /* Get InodeTable field from the ExtGroupDescriptor structure */
+    mov eax, dword ptr es:[si + 8]
+
+    /* Sum EAX with Inode index block */
+    add eax, dword ptr BP_REL(ExtReadInodeIndexBlock)
+
+    jmp ExtReadBlock
+
+msgDiskError:
+    .ascii "Disk error", CR, LF, NUL
+msgAnyKey:
+    .ascii "Press key", CR, LF, NUL
+
+    .org 509
+
+BootPartition:
+    .byte 0
+
+    .word HEX(0aa55)       // BootSector signature
+
+// End of bootsector
+//
+// Now starts the extra boot code that we will store
+// at sector 1 on a EXT2 volume
+
+LoadRootDirectory:
+    xor eax, eax
+    mov al, EXT_ROOT_INO			// Put the root directory inode number in EAX
+    call ExtReadInode				// Read the inode
+
+    call ReadDirectPointers         // Read the direct pointers
+    call FindFile                   // Find the extended loader and run it
+
+    jmp ExtLdrPrintFileNotFound
+
+ReadDirectPointers:
+    push es
+    pushad
+    xor eax, eax                                        // Clear eax register because it will be used for multiplication
+    call ExtSetInodeSegment
+
+    mov si, word ptr BP_REL(ExtReadInodeIndexOffset)
+
+    mov dword ptr BP_REL(ExtFileAddress), FREELDR_BASE
+
+    /* Calculate direct pointers amount with block size and enforce limits */
+    mov al, EXT_INODE_BLOCKS
+    mul dword ptr BP_REL(ExtBlockSizeInBytes)
+
+    /* Load file size from memory */
+    mov ebx, dword ptr es:[si + 4]
+
+    /* Compare and limit file size */
+    cmp ebx, eax
+    jbe .BelowOrEqualSize
+    mov ebx, eax
+.BelowOrEqualSize:
+    mov dword ptr BP_REL(ExtFileSize), ebx
+
+    /* Set rounded up file size */
+    add ebx, dword ptr BP_REL(ExtBlockSizeInBytes)
+    dec ebx
+    mov dword ptr BP_REL(ExtFileSizeState), ebx
+
+    xor ecx, ecx
+    mov cl, EXT_INODE_BLOCKS
+    add si, 40
+.LLoop:
+    call ExtSetInodeSegment
+
+    mov eax, dword ptr BP_REL(ExtFileSizeState)
+    cmp eax, dword ptr BP_REL(ExtBlockSizeInBytes)
+    jb .LDone
+
+    mov eax, dword ptr es:[si]
+
+    call ExtSetFileSegment
+
+    test eax, eax
+    jz .LFillBlock
+
+    call ExtReadBlock
+    jmp .LLoopEnd
+.LFillBlock:
+    xor di, di
+    xor ax, ax
+    mov cx, word ptr BP_REL(ExtBlockSizeInBytes)
+    rep stosb
+.LLoopEnd:
+    mov eax, dword ptr BP_REL(ExtBlockSizeInBytes)
+    sub dword ptr BP_REL(ExtFileSizeState), eax
+    add dword ptr BP_REL(ExtFileAddress), eax
+
+    add si, 4
+    loop .LLoop
+.LDone:
+    mov dword ptr BP_REL(ExtFileAddress), FREELDR_BASE
+    popad
+    pop es
+    ret
+
+FindFile:
+    call ExtSetFileSegment
+    call SwapESWithDS
```

Comment:
Do you gain any bytes by doing a call to this helper, instead of inlining what it does?